File: challenger.R
data = read.csv('challenger-data.csv')

install.packages('plyr')
library(plyr)

failures = subset(data, data$Y == 1)
success = subset(data, data$Y == 0)
failures_freq = count(failures, 'X')
success_freq = count(success, 'X')
plot(success$X, integer(length(success$X)), ylim=c(-0.5, 3.5), col='blue', xlab='X: Temperature', ylab='Number of failures', pch=19)
points(failures_freq$X, failures_freq$freq, col='red', pch=19)

model = glm(data$Y ~ data$X, family=binomial(link='logit'), data=data)
summary(model)

File: challenger.py
from matplotlib import pyplot as plt
from patsy import dmatrices
import matplotlib as mpl
import numpy as np
import pandas as pd
import statsmodels.discrete.discrete_model as sm

class Challenger:
    """Loaded data from csv file"""
    data = []

    def __init__(self):
        self.data = pd.read_csv("./challenger-data.csv")

    def plot(self):
        failures = self.data.loc[(self.data.Y == 1)]
        success = self.data.loc[(self.data.Y == 0)]
        failures_freq = failures.X.value_counts()
        success_freq = success.X.value_counts()
        plt.scatter(failures_freq.index, failures_freq, c='red', s=40)
        plt.scatter(success_freq.index, np.zeros(len(success_freq)), c='blue', s=40)
        plt.xlabel('X: Temperature')
        plt.ylabel('Number of failures')
        plt.show()

    def model(self):
        y, X = dmatrices('Y~X', self.data, return_type='dataframe')
        logit = sm.Logit(y, X)
        result = logit.fit()
        print result.summary()

    def __main__(self):
        self.model();

if __name__ == '__main__':
    challenger = Challenger()
    challenger.model()
